### PR TITLE
[Unity integration] Ably UI console example app

### LINKED ISSUE #1135: [Unity-Integration] Add UI Example
- Add a working UI example for the ably unity library.
- Include it as a part of export `.unitypackage`

┆Issue is synchronized with this [Jira Uncategorised](https://ably.atlassian.net/browse/SDK-1657) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by sacOO7**
- Created Ably UI console app.
- Added basic functionality for showing active connection with the server.
- Set `AblyConsole -> _apiKey` to get example working

1. No connect 

Ably No Connect

2. Connected

ably_connected

3. Messages

ably_messages

4. Disconnected

disconnected

- Fixes #1135

### CODE REVIEW

**Comment by tomkirbygreen** [comment]
This looks great Sachin. Looking forward to getting it running on my VM and then I'll review the code. Thank you! :-)

**Comment by tomkirbygreen** [comment]
Could we please change the `Example` folder to be a little more specific and allow for future examples. So I would suggest changing `unity/Assets/Ably/Example/` to `unity/Assets/Ably/Examples/Chat/`. Thank you.

File: unity/Assets/Ably/Example/AblyChannelUiConsole.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            _publish = GameObject.Find("Publish").GetComponent<Button>();
+            _publish.onClick.AddListener(PublishClickHandler);
+
+            _channelName = GameObject.Find("ChannelName").GetComponent<InputField>();
+            _eventName = GameObject.Find("EventName").GetComponent<InputField>();
+            _payload = GameObject.Find("Payload").GetComponent<InputField>();
+            EnableUiComponents(false);
+        }
+
+
```
Comment (tomkirbygreen) [inline unity/Assets/Ably/Example/AblyChannelUiConsole.cs:56]: Single line between functions please.

File: unity/Assets/Ably/Example/AblyConsole.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    {
+        private AblyRealtime _ably;
+        private ClientOptions _clientOptions;
+
+        private Text _textContent;
+        private InputField _clientId;
+        private Button _connectButton;
+        private Button _connectionStatus;
+
+        private static string _apiKey = "";
```
Comment (tomkirbygreen) [inline unity/Assets/Ably/Example/AblyConsole.cs:20]: `= String.Empty;` please.

File: unity/Assets/Ably/Example/AblyConsole.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                _ablyChannelUiConsole.EnableUiComponents(_isConnected);
+                _ablyPresenceUiConsole.EnableUiComponents(_isConnected);
+                _connectButton.GetComponentInChildren<Text>().text = _isConnected ? "Disconnect" : "Connect";
+            });
+        }
+
+        private void ConnectClickHandler()
+        {
+            _clientOptions.ClientId = _clientId.text;
+            if (_isConnected)
```
Comment (tomkirbygreen) [inline unity/Assets/Ably/Example/AblyConsole.cs:100]: Missing `{...}` for `if` and `else` blocks.

File: unity/Assets/Ably/Example/AblyConsole.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            Debug.Log(message);
+            _textContent.text = $"{_textContent.text}\n{message}";
+        }
+    }
+
+    internal interface IUiConsole
+    {
+        void LogAndDisplay(string message);
+    }
+}
```
Comment (tomkirbygreen) [inline unity/Assets/Ably/Example/AblyConsole.cs:117]: Missing newline.

File: unity/Assets/Ably/Example/AblyChannelUiConsole.cs [hunk 5]
```diff
@@ -0,0 +1,125 @@
+﻿using System.Linq;
+using IO.Ably;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Assets.Ably.Example
```
Comment (tomkirbygreen) [inline unity/Assets/Ably/Example/AblyChannelUiConsole.cs:6]: This namespace needs to change to `namespace Asserts.Ably.Examples.Chat` in line with the comment about renaming the project from `Example` to `Examples/Chat`.

**Review by tomkirbygreen — CHANGES_REQUESTED** [review]
Just a few action points.

File: unity/Assets/Ably/Example/AblyChannelUiConsole.cs (hunk 1 above)
Comment (sacOO7) [reply unity/Assets/Ably/Example/AblyChannelUiConsole.cs]: Done

File: unity/Assets/Ably/Example/AblyConsole.cs (hunk 2 above)
Comment (sacOO7) [reply unity/Assets/Ably/Example/AblyConsole.cs]: Updated to `Your_Api-key_Here`

File: unity/Assets/Ably/Example/AblyConsole.cs (hunk 3 above)
Comment (sacOO7) [reply unity/Assets/Ably/Example/AblyConsole.cs]: done

File: unity/Assets/Ably/Example/AblyChannelUiConsole.cs (hunk 5 above)
Comment (sacOO7) [reply unity/Assets/Ably/Example/AblyChannelUiConsole.cs]: done

File: unity/Assets/Ably/Example/AblyConsole.cs (hunk 4 above)
Comment (sacOO7) [reply unity/Assets/Ably/Example/AblyConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                _uiConsole.LogAndDisplay($"Received message <b>{message.Data}</b> from channel <b>{channelName}</b>");
+            });
+            _uiConsole.LogAndDisplay($"Successfully subscribed to channel <b>{channelName}</b>");
+        }
+
+        private void UnsubscribeChannelHandler()
+        {
+            var channelName = _channelName.text;
+            _ably.Channels.Get(channelName).Unsubscribe();
+            _uiConsole.LogAndDisplay($"Successfully unsubscribed to channel <b>{channelName}</b>");
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs:71]:
```suggestion
            _uiConsole.LogAndDisplay($"Successfully unsubscribed from channel <b>{channelName}</b>");
```

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            var channelName = _channelName.text;
+            var eventName = _eventName.text;
+            _ably.Channels.Get(channelName).Subscribe(eventName, message =>
+            {
+                _uiConsole.LogAndDisplay($"Received message <b>{message.Data}</b> from channel <b>{channelName}</b>");
+            });
+            _uiConsole.LogAndDisplay($"Successfully subscribed to channel <b>{channelName}</b>");
+        }
+
+        private void UnsubscribeChannelHandler()
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs:67]: Not sure if naming methods ending with -Handler is a common practice, but this method name sounds more like an object. Wouldn't it be cleaner if it was `UnsubscribeFromChannel`?

Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs:67]:
```suggestion
        private void UnsubscribeFromChannel()
```

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+            _uiConsole.LogAndDisplay($"Successfully unsubscribed to channel <b>{channelName}</b>");
+        }
+
+        private void ListChannelsHandler()
+        {
+            var channelNames = string.Join(", ", _ably.Channels.Select(channel => channel.Name));
+            _uiConsole.LogAndDisplay($"Channel Names - <b>{channelNames}</b>");
+        }
+
+        private async void ChannelMessageHistoryHandler()
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs:80]:
```suggestion
        private async void LoadChannelMessageHistory()
```

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                if (historyPage.IsLast)
+                {
+                    break;
+                }
+                historyPage = await historyPage.NextAsync();
+            };
+            _uiConsole.LogAndDisplay($"#### <b>{channelName}</b> ####");
+        }
+
+        private async void PublishClickHandler()
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs:100]:
```suggestion
        private async void PublishMessage()
```

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+            _leavePresence = GameObject.Find("LeavePresence").GetComponent<Button>();
+            _leavePresence.onClick.AddListener(PresenceLeaveClickHandler);
+
+            _channelName = GameObject.Find("PresenceChannel").GetComponent<InputField>();
+            _payload = GameObject.Find("PresencePayload").GetComponent<InputField>();
+            EnableUiComponents(false);
+        }
+
+        internal void PresenceSubscribeClickHandler()
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs:57]:
```suggestion
        internal void SubscribeToPresence()
```

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            var channelName = _channelName.text;
+            _ably.Channels.Get(channelName).Presence.Subscribe(message =>
+            {
+                _uiConsole.LogAndDisplay($"Received presence message <b>{message.Data}</b> from channel <b>{channelName}</b> with clientId <b>{message.ClientId}</b>");
+            });
+            _uiConsole.LogAndDisplay($"Successfully subscribed to channel <b>{channelName}</b> for <b>Presence</b> messages");
+        }
+
+        internal void PresenceUnsubscribeClickHandler()
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs:67]:
```suggestion
        internal void UnsubscribeFromPresence()
```

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            var channelName = _channelName.text;
+            _ably.Channels.Get(channelName).Presence.Unsubscribe();
+            _uiConsole.LogAndDisplay($"Successfully unsubscribed to channel <b>{channelName}</b> for <b>Presence</b> messages");
+        }
+
+        internal async void GetPresenceClickHandler()
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs:74]:
```suggestion
        internal async void GetPresence()
```

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+            var presenceMessages = await _ably.Channels.Get(channelName).Presence.GetAsync();
+            _uiConsole.LogAndDisplay($"#### <b>{channelName}</b> Presence ####");
+            foreach (var presenceMessage in presenceMessages)
+            {
+                _uiConsole.LogAndDisplay($"ClientID - <b>{presenceMessage.ClientId}</b>, Message - <b>{presenceMessage.Data}</b>");
+            }
+            _uiConsole.LogAndDisplay($"#### <b>{channelName}</b> ####");
+        }
+
+        internal async void PresenceMessageHistoryClickHandler()
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs:86]:
```suggestion
        internal async void GetPresenceMessageHistory()
```

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+            // async-await makes sure call is executed in the background and then result is posted on UnitySynchronizationContext/Main thread
+            var result = await _ably.Channels.Get(channelName).Presence.EnterAsync(_payload.text);
+
+            _uiConsole.LogAndDisplay(result.IsSuccess
+                ? $"Successfully entered presence to channel <b>{channelName}</b> for clientID <b>{_ably.ClientId}</b>"
+                : $"Error entering presence to channel <b>{channelName}</b>, failed with error <b>{result.Error.Message}</b>");
+        }
+
+        internal async void PresenceLeaveClickHandler()
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs:118]:
```suggestion
        internal async void LeavePresence()
```

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            var channelName = _channelName.text;
+
+            // async-await makes sure call is executed in the background and then result is posted on UnitySynchronizationContext/Main thread
+            var result = await _ably.Channels.Get(channelName).Presence.LeaveAsync(_payload.text);
+
+            _uiConsole.LogAndDisplay(result.IsSuccess
+                ? $"Successfully left presence to channel <b>{channelName}</b> for clientID <b>{_ably.ClientId}</b>"
+                : $"Error leaving presence to channel <b>{channelName}</b>, failed with error <b>{result.Error.Message}</b>");
```
Comment (ikbalkaya) [inline unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs:127]:
```suggestion
                : $"Error leaving presence from channel <b>{channelName}</b>, failed with error <b>{result.Error.Message}</b>");
```

**Review by ikbalkaya — CHANGES_REQUESTED** [review]

File: unity/Assets/Ably/Examples/Chat/AblyConsole.cs [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+                _ably.Connect();
+            }
+        }
+
+        public void LogAndDisplay(string message)
+        {
+            Debug.Log(message);
+            _textContent.text = $"{_textContent.text}\n{message}";
+        }
+
```
Comment (tomkirbygreen) [inline unity/Assets/Ably/Examples/Chat/AblyConsole.cs:116]: Remove please.

**Review by tomkirbygreen — APPROVED** [review]
The actual logic is fine I think. There are some 'fit and finish' issues, but I don't know if they should hold things up.

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs (hunk 6 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs (hunk 7 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+            _publish = GameObject.Find("Publish").GetComponent<Button>();
+            _publish.onClick.AddListener(PublishClickHandler);
+
+            _channelName = GameObject.Find("ChannelName").GetComponent<InputField>();
+            _eventName = GameObject.Find("EventName").GetComponent<InputField>();
+            _payload = GameObject.Find("Payload").GetComponent<InputField>();
+            EnableUiComponents(false);
+        }
+
+        private void SubscribeChannelClickHandler()
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs (hunk 8 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs (hunk 9 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyChannelUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs (hunk 10 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs (hunk 11 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+                _uiConsole.LogAndDisplay($"Received presence message <b>{message.Data}</b> from channel <b>{channelName}</b> with clientId <b>{message.ClientId}</b>");
+            });
+            _uiConsole.LogAndDisplay($"Successfully subscribed to channel <b>{channelName}</b> for <b>Presence</b> messages");
+        }
+
+        internal void PresenceUnsubscribeClickHandler()
+        {
+            var channelName = _channelName.text;
+            _ably.Channels.Get(channelName).Presence.Unsubscribe();
+            _uiConsole.LogAndDisplay($"Successfully unsubscribed to channel <b>{channelName}</b> for <b>Presence</b> messages");
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs (hunk 12 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs (hunk 13 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+                if (presenceHistoryPage.IsLast)
+                {
+                    break;
+                }
+                presenceHistoryPage = await presenceHistoryPage.NextAsync();
+            };
+            _uiConsole.LogAndDisplay($"#### <b>{channelName}</b> ####");
+        }
+
+        internal async void PresenceEnterClickHandler()
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs (hunk 14 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs]: done

File: unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs (hunk 15 above)
Comment (sacOO7) [reply unity/Assets/Ably/Examples/Chat/AblyPresenceUiConsole.cs:127]: done

**Comment by tomkirbygreen** [comment]
Looking good @sacOO7 , we'll just need to wait for @ikbalkaya to approve it assuming he's happy now.

**Review by ikbalkaya — APPROVED** [review]
LGTM

**Pull request merged**